src/idl/src/annotation.c

```diff
@@ -174,7 +174,7 @@ annotate_optional(
     idl_error(pstate, idl_location(annotation_appl),
       "@optional can only be assigned to members");
     return IDL_RETCODE_SEMANTIC_ERROR;
-  } else if (mem->key.value) {
+  } else if (value && mem->key.value) {
```

Comment:
You are correct.

It would be even better if the value (true/false) of the annotation would be read before this statement, since if it is false (which would mean the same as no `@optional` annotation), the @optional annotation will have no possible conflict with an `@key`, if any.